src/transformations.jl

```diff
@@ -97,22 +143,46 @@ julia> StatsBase.transform(dt, X)
  -1.0   0.0  1.0
 ```
 """
-function fit(::Type{ZScoreTransform}, X::AbstractArray{<:Real,2}; center::Bool=true, scale::Bool=true)
-    d, n = size(X)
-    n >= 2 || error("X must contain at least two columns.")
-
+function fit(::Type{ZScoreTransform}, X::AbstractMatrix{<:Real};
+             dims::Union{Integer,Nothing}=nothing, center::Bool=true, scale::Bool=true)
+    if dims == 1
+        n, l = size(X)
+        n >= 2 || error("X must contain at least two rows.")
+        m, s = mean_and_std(X, 1)
+    elseif dims == 2
+        l, n = size(X)
+        n >= 2 || error("X must contain at least two columns.")
+        m, s = mean_and_std(X, 2)
+    elseif dims === nothing
+        Base.depwarn("fit(t, x) is deprecated: use fit(t, x, dims=2) instead", :fit)
+        m, s = mean_and_std(X, 2)
+    else
+        throw(DomainError(dims, "fit only accept dims to be 1 or 2."))
+    end
     T = eltype(X)
-    m, s = mean_and_std(X, 2)
+    return ZScoreTransform(l, dims, (center ? vec(m) : zeros(T, 0)),
+                                    (scale ? vec(s) : zeros(T, 0)))
+end
+
+function fit(::Type{ZScoreTransform}, X::AbstractVector{<:Real};
+             dims::Union{Integer,Nothing}=nothing, center::Bool=true, scale::Bool=true)
+    if dims == nothing
+        Base.depwarn("fit(t, x) is deprecated: use fit(t, x, dims=2) instead", :fit)
+    elseif dims != 1
+        throw(DomainError(dims, "fit only accept dims=1 over a vector. Try fit(t, x, dims=1)."))
+    end
 
-    return ZScoreTransform(d, (center ? vec(m) : zeros(T, 0)),
-                              (scale ? vec(s) : zeros(T, 0)))
+    T = eltype(X)
+    m, s = mean_and_std(X)
+    return ZScoreTransform(1, dims, (center ? [m] : zeros(T, 0)),
+                                    (scale ? [s] : zeros(T, 0)))
 end
 
-function transform!(y::AbstractVecOrMat{<:Real}, t::ZScoreTransform, x::AbstractVecOrMat{<:Real})
-    d = t.dim
-    size(x,1) == size(y,1) == d || throw(DimensionMismatch("Inconsistent dimensions."))
-    n = size(y,2)
-    size(x,2) == n || throw(DimensionMismatch("Inconsistent dimensions."))
+function transform!(y::AbstractMatrix{<:Real}, t::ZScoreTransform, x::AbstractMatrix{<:Real})
```

Comment:
This function doesn't take into account `t.dims`? Same for the other one below.

Reply:
I move the `t.dims` check above (i.e. `transform!(t::AbstractDataTransform, x::AbstractMatrix{<:Real})` and related `transform`, `reconstruct!`, `reconstruct`).
Here, it would focus on dealing with column-major calculation.

Reply:
But that method is publicly available, we can't assume people will call it and expect the column-major convention.

Reply:
Maybe I didn't explain myself clearly. Those methods are publicly available. Users call those methods directly, and `t.dims` is specified. In that, further operations are dispatched to internals (i.e. `transform!(x, t, x)`) by `t.dims`. There is the only one version `transform!(x, t, x)` operates in column-major convention for performance. Therefore, users are allowed to use the operations they want, and the internals just conduct in column-major convention.

Reply:
@nalimilan is correct. You need to move `t.dim` checks from `transfrom` to `transform!`.

Reply:
OK, I see it. I didn't notice that `transform!(y, t, x)` is also public.

Reply:
Do we need to public `transform!(y, t, x)`? It's not tested and lack of documentation, but public before.

Reply:
`transform!` is public

Reply:
OK.